src/js/components/PlacementConstraintsPartial.js

```diff
@@ -0,0 +1,238 @@
+import React, { Component } from "react";
+import { Tooltip } from "reactjs-components";
+
+import { findNestedPropertyInObject } from "#SRC/js/utils/Util";
+import AddButton from "#SRC/js/components/form/AddButton";
+import DeleteRowButton from "#SRC/js/components/form/DeleteRowButton";
+import FieldAutofocus from "#SRC/js/components/form/FieldAutofocus";
+import FieldError from "#SRC/js/components/form/FieldError";
+import FieldHelp from "#SRC/js/components/form/FieldHelp";
+import FieldInput from "#SRC/js/components/form/FieldInput";
+import FieldLabel from "#SRC/js/components/form/FieldLabel";
+import FieldSelect from "#SRC/js/components/form/FieldSelect";
+import FormGroup from "#SRC/js/components/form/FormGroup";
+import FormGroupHeading from "#SRC/js/components/form/FormGroupHeading";
+import FormGroupHeadingContent
+  from "#SRC/js/components/form/FormGroupHeadingContent";
+import FormRow from "#SRC/js/components/form/FormRow";
+import Icon from "#SRC/js/components/Icon";
+
+import OperatorTypes from "#PLUGINS/services/src/js/constants/OperatorTypes";
+import PlacementConstraintsUtil
+  from "#PLUGINS/services/src/js/utils/PlacementConstraintsUtil";
+
+function placementConstraintLabel(name, tooltipText, options = {}) {
+  const { isRequired = false, linkText = "More information" } = options;
+
+  const tooltipContent = (
+    <span>
+      {`${tooltipText} `}
+      <a
+        href="https://mesosphere.github.io/marathon/docs/constraints.html"
+        target="_blank"
+      >
+        {linkText}
+      </a>.
+    </span>
+  );
+
+  return (
+    <FieldLabel>
+      <FormGroupHeading required={isRequired}>
+        <FormGroupHeadingContent primary={true}>
+          {name}
+        </FormGroupHeadingContent>
+        <FormGroupHeadingContent>
+          <Tooltip
+            content={tooltipContent}
+            interactive={true}
+            maxWidth={300}
+            wrapText={true}
+          >
+            <Icon color="grey" id="circle-question" size="mini" />
+          </Tooltip>
+        </FormGroupHeadingContent>
+      </FormGroupHeading>
+    </FieldLabel>
+  );
+}
+
+class PlacementSection extends Component {
+  getOperatorTypes() {
+    return Object.keys(OperatorTypes).map((type, index) => {
+      return <option key={index} value={type}>{type}</option>;
+    });
+  }
+
+  getPlacementConstraintsFields(data = []) {
+    const constraintsErrors = findNestedPropertyInObject(
+      this.props.errors,
+      "constraints"
+    );
+    const hasOneRequiredValue = data.some(function(constraint) {
+      return PlacementConstraintsUtil.requiresValue(constraint.operator);
+    });
+    const hideValueColumn = data.every(function(constraint) {
+      return PlacementConstraintsUtil.requiresEmptyValue(constraint.operator);
+    });
+
+    return data.map((constraint, index) => {
```

Comment:
Please sync with @Poltergeist regarding this `index`. To implement the filtering we need, we have to stop relying on the map-index, as the `data` array could be pre-filtered thus, the native array index won't reflect the position of the constraint in the app definition JSON.

Reply:
Yes because of some constraints which should be displayed outside of the normal form we can't depend on the index of the array being right.

So as an example we do have the following example in the JSON.

```JS
let constraints = [ 
  ["@special", "IS", "SOMETHING"], 
  ["hostname" "UNIQUE"]
]
```
The special value we do not want to show meaning we would filter it before handing it off, But we really need to have the right index on the hostname.

So the structure I would like to propose would be:
```JS
let constraints = [
  {
    constraint: ["hostname", "UNIQUE"],
    index: 1
  }
]
```

Reply:
Did we merge @Poltergeist PR?

Reply:
as discussed: Refactoring later, its mergable.